core/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java

```diff
@@ -100,43 +118,64 @@ private void updateLastActivity(PotentialConflictDTO payload, Date activityEndTi
 				.collect(Collectors.toSet());
 	}
 
-	private void sendConflictMessageToAllDestinationsOfUser(PotentialConflictDTO payload, UserAnonymizedDTO userAnonymized,
+	private void sendConflictMessageToAllDestinationsOfUser(ActivityPayload payload, UserAnonymizedDTO userAnonymized,
 			Activity activity)
 	{
-		GoalConflictMessage selfGoalConflictMessage = sendConflictMessage(payload, activity,
+		GoalConflictMessage selfGoalConflictMessage = sendConflictMessage(payload, userAnonymized, activity,
 				userAnonymized.getAnonymousDestination(), null);
 
 		userAnonymized.getBuddyDestinations().stream()
-				.forEach(d -> sendConflictMessage(payload, activity, d, selfGoalConflictMessage));
+				.forEach(d -> sendConflictMessage(payload, userAnonymized, activity, d, selfGoalConflictMessage));
 	}
 
-	private GoalConflictMessage sendConflictMessage(PotentialConflictDTO payload, Activity activity,
+	private GoalConflictMessage sendConflictMessage(ActivityPayload payload, UserAnonymizedDTO userAnonymized, Activity activity,
 			MessageDestinationDTO destination, GoalConflictMessage origin)
 	{
 		GoalConflictMessage message;
 		if (origin == null)
 		{
-			message = GoalConflictMessage.createInstance(payload.getVPNLoginID(), activity, payload.getURL());
+			message = GoalConflictMessage.createInstance(userAnonymized.getID(), activity, payload.url);
 		}
 		else
 		{
-			message = GoalConflictMessage.createInstanceFromBuddy(payload.getVPNLoginID(), origin);
+			message = GoalConflictMessage.createInstanceFromBuddy(userAnonymized.getID(), origin);
 		}
 		messageService.sendMessage(message, destination);
 		return message;
 	}
 
-	private Set<Goal> determineMatchingGoalsForUser(UserAnonymizedDTO userAnonymized, Set<String> categories)
+	private Set<Goal> determineMatchingGoalsForUser(UserAnonymizedDTO userAnonymized,
+			Set<ActivityCategoryDTO> matchingActivityCategories)
 	{
-		Set<ActivityCategoryDTO> allActivityCategories = activityCategoryService.getAllActivityCategories();
-		Set<UUID> matchingActivityCategoryIDs = allActivityCategories.stream().filter(ac -> {
-			Set<String> acSmoothwallCategories = new HashSet<>(ac.getSmoothwallCategories());
-			acSmoothwallCategories.retainAll(categories);
-			return !acSmoothwallCategories.isEmpty();
-		}).map(ac -> ac.getID()).collect(Collectors.toSet());
+		Set<UUID> matchingActivityCategoryIDs = matchingActivityCategories.stream().map(ac -> ac.getID())
+				.collect(Collectors.toSet());
 		Set<Goal> goalsOfUser = userAnonymized.getGoals();
 		Set<Goal> matchingGoalsOfUser = goalsOfUser.stream()
 				.filter(g -> matchingActivityCategoryIDs.contains(g.getActivityCategory().getID())).collect(Collectors.toSet());
 		return matchingGoalsOfUser;
 	}
+
+	class ActivityPayload
+	{
+		public String url;
```

Comment:
Using a struct-like approach with all public member variables is fine, but then make all members final, so they cannot be modified after creation.